File: frontend/src/games/core/seeded-random.ts
export class SeededRandom {
  private state: number;

  constructor(seed: number) {
    this.state = (seed >>> 0) || 0x6d2b79f5;
  }

  nextIndex(bound: number) {
    if (!Number.isInteger(bound) || bound <= 0) {
      throw new Error("Bound must be a positive integer");
    }
    let value = this.state;
    value ^= value << 13;
    value ^= value >>> 17;
    value ^= value << 5;
    this.state = value >>> 0;
    return Math.floor((this.state * bound) / 4_294_967_296);
  }
}

export function randomSeed() {
  const values = new Uint32Array(1);
  window.crypto.getRandomValues(values);
  return values[0] || 1;
}

File: backend/src/main/java/com/gameio/gameresult/replay/SeededRandom.java
package com.gameio.gameresult.replay;

public final class SeededRandom {
    private int state;

    public SeededRandom(long seed) {
        state = (int) seed;
        if (state == 0) {
            state = 0x6d2b79f5;
        }
    }

    public int nextIndex(int bound) {
        if (bound <= 0) {
            throw new IllegalArgumentException("Bound must be positive");
        }
        int value = state;
        value ^= value << 13;
        value ^= value >>> 17;
        value ^= value << 5;
        state = value;
        long unsigned = Integer.toUnsignedLong(state);
        return (int) ((unsigned * bound) / 4_294_967_296L);
    }
}
